apps/linux/vim.talon

```diff
@@ -0,0 +1,799 @@
+
+action(edit.extend_line_end):
+    user.vim_visual_mode("$")
+action(edit.extend_left):
+    user.vim_visual_mode("h")
+action(edit.extend_right):
+    user.vim_visual_mode("l")
+action(edit.extend_line_up):
+    user.vim_visual_mode("k")
+action(edit.extend_line_down):
+    user.vim_visual_mode("j")
+action(edit.extend_word_left):
+    user.vim_visual_mode("b")
+action(edit.extend_word_right):
+    user.vim_visual_mode("w")
+action(edit.extend_line_start):
+    user.vim_visual_mode("^")
+action(edit.extend_file_start):
+    user.vim_visual_mode("gg")
+action(edit.extend_file_end):
+    user.vim_visual_mode("G")
+
+action(edit.indent_more):
+    user.vim_normal_mode(">>")
+action(edit.indent_less):
+    user.vim_normal_mode("<<")
+action(edit.delete_line):
+    user.vim_normal_mode("dd")
+action(edit.delete):
+    user.vim_normal_mode_key(x)
+
+# note these are for mouse highlighted copy/paste. shouldn't be used for actual
+# vim commands
+action(edit.copy):
+    key(ctrl-shift-c)
+action(edit.paste):
+    key(ctrl-shift-v)
+
+###
+# `code/vim.py` actions based on vimspeak
+###
+# commands that can be triggered in visual or normal mode, and generally don't
+# have counting, etc
+<user.vim_normal_counted_motion_command>:
+    insert("{vim_normal_counted_motion_command}")
+#<user.vim_counted_command>:
+#    insert("{vim_counted_command}")
+<user.vim_motions_all_adjust>:
+    insert("{vim_motions_all_adjust}")
+# XXX - needs to be adjusted to handle the actions associated with key presses
+<user.vim_normal_counted_action>:
+    insert("{vim_normal_counted_action}")
+
+###
+# File editing and management
+###
+# NOTE: using `save` alone conflicts too much with the `say`
+save file:
+    user.vim_command_mode(":w\n")
+save [file] as:
+    key(escape)
+    user.vim_normal_mode(":w ")
+save all:
+    user.vim_normal_mode_exterm(":wa\n")
+save and (quit|close):
+    user.vim_normal_mode(":wq\n")
+(close|quit) file:
+    user.vim_normal_mode(":q\n")
+
+# no \n as a saftey measure
+(close|quit) all:
+    user.vim_normal_mode_exterm(":qa")
+
+force (close|quit) all:
+    user.vim_normal_mode_exterm(":qa!")
+
+force (close|quit):
+    user.vim_normal_mode_exterm(":q!\n")
+refresh file:
+    user.vim_normal_mode(":e!\n")
+edit [file|new]:
+    user.vim_normal_mode_exterm(":e ")
+reload [vim] config:
+    user.vim_normal_mode_exterm(":so $MYVIMRC\n")
+
+# For when the VIM cursor is hovering on a path
+open [this] link: user.vim_normal_mode("gx")
+open this file: user.vim_normal_mode("gf")
+open this file in [split|window]:
+    user.vim_set_normal_mode()
+    key(ctrl-w)
+    key(f)
+open this file in vertical [split|window]:
+    user.vim_normal_mode(":vertical wincmd f\n")
+
+(show|list) current directory: user.vim_normal_mode(":pwd\n")
+change buffer directory: user.vim_normal_mode(":lcd %:p:h\n")
+
+###
+# Standard commands
+###
+# XXX - technically should be handled by vim.py, but need to add "ctrl" key support
+redo:
```

Comment:
I'd suggest implementing the `edit.redo()` action instead, so we can stick with the global command 

`redo that: edit.redo()`

Reply:
this is actually change in the latest unpushed to this PR version so its already part of `vim.py`. I think will just support both. 

The main reason I want it part of `vim.py` as that it's one of the main action commands that is countable in `vim.py`, and the counting works differently in talon itself (count prefix, not ordinal suffix) but as a vim user it feels very natural to do it the vim way. it also allows vim to repeat counted commands with the `repeat` command, or record into macros . strictly speaking i suspect redo and undo are unlikely to get used in macros so they don't have to be counted this way, but because when you're heavily using vim all the other commands you end up counting the same way, suddenly not being able to count redo and undo I found quite unintuitive which is also why I added them to `vim.py` eventually. 

all that said having the `redo that` and `undo that` commands  supported in general via `actions` makes sense, but still also supporting the way I'm currently doing it will be the best of both worlds.

Reply:
Makes sense